### DISCUSSION TITLE
Best practice for global usage in server and client components

### DISCUSSION THREAD

**Discussion by pschaub** [amannn/next-intl#405] [Q&A] [answered]
### Is your feature request related to a problem? Please describe.

Currently I use `next-intl` with app router in [server components](https://next-intl-docs.vercel.app/docs/getting-started/app-router-server-components) and would like to use them in [client components](https://next-intl-docs.vercel.app/docs/getting-started/app-router-client-components) too.

In the [documentation for client components](https://next-intl-docs.vercel.app/docs/getting-started/app-router-client-components) it explains that I need to use the `NextIntlClientProvider`:
```ts
import {NextIntlClientProvider} from 'next-intl';
import {notFound} from 'next/navigation';
 
export function generateStaticParams() {
  return [{locale: 'en'}, {locale: 'de'}];
}
 
export default async function LocaleLayout({children, params: {locale}}) {
  let messages;
  try {
    messages = (await import(`../../messages/${locale}.json`)).default;
  } catch (error) {
    notFound();
  }
 
  return (
    <html lang={locale}>
      <body>
        <NextIntlClientProvider locale={locale} messages={messages}>
          {children}
        </NextIntlClientProvider>
      </body>
    </html>
  );
}
```

In this example the `locale` and `messages` are statically set in `app/[locale]/layout.tsx`.

**But because I use server components I have already set these values in `i18n.ts` and `middleware.ts`.**

### Describe the solution you'd like

I would like to know the best practice to integrate a combined global usage of `next-intl` for server components and client components with app router. It sometimes happens that you have server components mixed with client components.

**Comment by amannn** [comment, marked as answer]
Thank you for the question!

Did you find the page in the docs that goes into depth about [internationalization of Server & Client Comonents](https://next-intl-docs.vercel.app/docs/environments/server-client-components)? The page tries to answer the question you're asking.



If you find something in the docs that is missing, please let me know! E.g. currently there's a link to the mentioned page in the next steps in the [getting started with Server Components page](https://next-intl-docs.vercel.app/docs/getting-started/app-router-server-components#applocalepagetsx), but maybe it's a bit subtle.

Generally, as soon as we have a stable release of the Server Components integration, I'd like to remove the getting started with Client Components page, therefore the page mentioned above will become the canonical source for learning about transitioning to the client side.

Hope this helps! I'll convert this to a discussion, since it's more a usage question than a feature request.

**Reply by pschaub** [reply]
I have not seen this additional documentation page. All my questions are answered there. Thank you very much!

**Reply by amannn** [reply]
That's great! I know the docs are a bit much currently, possibly it's easy to overlook the page. I hope that we can make the docs simpler again with the stable release of RSC support as mentioned above!

**Reply by chemicstry** [reply]
I was also having the same question, but the suggestions provided in the doc page do not seem to fully solve my problems. The `useMessages()` method can be used to obtain translations from the server configuration (`i18n.ts`), but I couldn't find any way to retrieve other properties (i.e. `defaultTranslationValues`). I assume everything could be reached through `useIntlContext()`, but it is not exposed to the public API. Are there any workarounds to have a single i18n configuration between server and client?

**Reply by amannn** [reply]
You can define a shared config in a module somewhere and import it into all places where it's needed (`i18n.ts`, a component rendering a provider, …).

On a side note, auto-inheriting config from `i18n.ts` is not implemented because:
1. You have less control over client bundle size
2. We could only inherit config from `i18n.ts` if `NextIntlClientProvider` is rendered in a Server Component (not when in Client Components)
3. `defaultTranslationValues` can't be passed to `NextIntlClientProvider` as props, because functions can't be serialized. Instead, they need to be imported into a Client Component, making them part of the module graph.

Hope this helps!
